File: src/test/java/org/galaxio/gatling/kafka/javaapi/examples/MatchSimulation.java
package org.galaxio.gatling.kafka.javaapi.examples;

import io.gatling.javaapi.core.ScenarioBuilder;
import io.gatling.javaapi.core.Simulation;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.galaxio.gatling.kafka.javaapi.KafkaDsl;
import org.galaxio.gatling.kafka.javaapi.protocol.KafkaProtocolBuilder;
import org.galaxio.gatling.kafka.request.KafkaProtocolMessage;

import java.time.Duration;
import java.util.Collections;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Supplier;
import java.util.stream.Stream;

import static io.gatling.javaapi.core.CoreDsl.atOnceUsers;
import static io.gatling.javaapi.core.CoreDsl.scenario;

public class MatchSimulation extends Simulation {

    private final KafkaProtocolBuilder kafkaProtocolMatchByValue = KafkaDsl.kafka()
            .producerSettings(
                    Map.of(
                            ProducerConfig.ACKS_CONFIG, "1",
                            ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092"
                    )
            )
            .consumeSettings(
                    Map.of("bootstrap.servers", "localhost:9092")
            )
            .timeout(Duration.ofSeconds(5))
            // for match by message value
            .matchByValue();

    private byte[] matchByOwnVal(KafkaProtocolMessage message) {
        // do something with the message and extract the values you are interested in
        // method is called:
        // - for each message which will be sent out
        // - for each message which has been received
        return "Custom Message".getBytes(); // just returning something
    }

    private final KafkaProtocolBuilder kafkaProtocolMatchByMessage = KafkaDsl.kafka()
            .producerSettings(
                    Map.of(
                            ProducerConfig.ACKS_CONFIG, "1",
                            ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092"
                    )
            )
            .consumeSettings(
                    Map.of(
                            "bootstrap.servers", "localhost:9092"
                    )
            )
            .timeout(Duration.ofSeconds(5))
            .matchByMessage(this::matchByOwnVal);

    private final AtomicInteger c = new AtomicInteger(0);
    private final Iterator<Map<String, Object>> feeder =
            Stream.generate((Supplier<Map<String, Object>>) () -> Collections.singletonMap("kekey", c.incrementAndGet())
            ).iterator();

    private final ScenarioBuilder scn = scenario("Basic")
            .feed(feeder)
            .exec(
                    KafkaDsl.kafka("ReqRep").requestReply()
                            .requestTopic("test.t")
                            .replyTopic("test.t")
                            .send("#{kekey}", """
                                    { "m": "dkf" }
                                    """));

    {
        setUp(
                scn.injectOpen(atOnceUsers(1)))
                .protocols(kafkaProtocolMatchByMessage)
                .maxDuration(Duration.ofSeconds(120));
    }

}

File: examples/kotlin/src/gatling/kotlin/org/galaxio/examples/kotlinapi/MatchSimulation.kt
package org.galaxio.examples.kotlinapi

import io.gatling.javaapi.core.CoreDsl.atOnceUsers
import io.gatling.javaapi.core.CoreDsl.global
import io.gatling.javaapi.core.CoreDsl.scenario
import io.gatling.javaapi.core.Simulation
import org.apache.kafka.clients.producer.ProducerConfig
import org.galaxio.gatling.kafka.javaapi.KafkaDsl.kafka
import org.galaxio.gatling.kafka.request.KafkaProtocolMessage
import java.time.Duration
import java.util.concurrent.atomic.AtomicInteger

class MatchSimulation : Simulation() {

    @Suppress("unused")
    private val kafkaProtocolMatchByValue = kafka()
        .producerSettings(
            mapOf<String, Any>(
                ProducerConfig.ACKS_CONFIG to "1",
                ProducerConfig.BOOTSTRAP_SERVERS_CONFIG to "localhost:9093",
            )
        )
        .consumeSettings(mapOf<String, Any>("bootstrap.servers" to "localhost:9093"))
        .timeout(Duration.ofSeconds(10))
        .matchByValue()

    private fun matchByOwnVal(message: KafkaProtocolMessage): ByteArray = "Custom Message".toByteArray()

    private val kafkaProtocolMatchByMessage = kafka()
        .producerSettings(
            mapOf<String, Any>(
                ProducerConfig.ACKS_CONFIG to "1",
                ProducerConfig.BOOTSTRAP_SERVERS_CONFIG to "localhost:9093",
            )
        )
        .consumeSettings(mapOf<String, Any>("bootstrap.servers" to "localhost:9093"))
        .timeout(Duration.ofSeconds(10))
        .matchByMessage { message: KafkaProtocolMessage -> matchByOwnVal(message) }

    private val c = AtomicInteger(0)

    private val feeder = generateSequence { mapOf<String, Any>("kekey" to c.incrementAndGet()) }.iterator()

    private val scn = scenario("Basic")
        .feed(feeder)
        .exec(
            kafka("ReqRep").requestReply()
                .requestTopic("ex.kotlin.match.t")
                .replyTopic("ex.kotlin.match.t")
                .send("#{kekey}", """{ "m": "dkf" }""")
        )

    init {
        setUp(scn.injectOpen(atOnceUsers(1)))
            .protocols(kafkaProtocolMatchByMessage)
            .maxDuration(Duration.ofSeconds(120))
            .assertions(
                global().allRequests().count().`is`(1L),
                global().successfulRequests().percent().`is`(100.0),
            )
    }
}
